istioctl/pkg/kubernetes/client.go

```diff
@@ -217,3 +220,62 @@ func (client *Client) GetPilotAgentContainer(podName, podNamespace string) (stri
 	}
 	return proxyContainer, nil
 }
+
+type podDetail struct {
+	binary    string
+	container string
+}
+
+// GetIstioVersions gets the version for each Istio component
+func (client *Client) GetIstioVersions(namespace string, short bool) (string, error) {
+	pods, err := client.GetIstioPods(namespace, map[string]string{"labelSelector": "istio"})
+	if err != nil {
+		return "", err
+	}
+	if len(pods) == 0 {
+		return "", errors.New("unable to find any Istio pod in namespace " + namespace)
+	}
+
+	labelToPodDetail := map[string]podDetail{
+		"pilot":            {"/usr/local/bin/pilot-discovery", "discovery"},
+		"citadel":          {"/usr/local/bin/istio_ca", "citadel"},
+		"egressgateway":    {"/usr/local/bin/pilot-agent", "egressgateway"},
+		"galley":           {"/usr/local/bin/galley", "validator"},
+		"ingressgateway":   {"/usr/local/bin/pilot-agent", "ingressgateway"},
+		"telemetry":        {"/usr/local/bin/mixs", "mixer"},
+		"policy":           {"/usr/local/bin/mixs", "mixer"},
+		"sidecar-injector": {"/usr/local/bin/sidecar-injector", "sidecar-injector-webhook"},
+	}
+
+	res := ""
+	for _, pod := range pods {
+		component := pod.Labels["istio"]
+
+		// Special cases
+		switch component {
+		case "statsd-prom-bridge":
+			continue
+		case "mixer":
+			component = pod.Labels["istio-mixer-type"]
+		}
+
+		res += fmt.Sprintf("Component: %s\n", component)
+		if detail, ok := labelToPodDetail[component]; ok {
+			cmd := []string{detail.binary, "version"}
```

Comment:
are you sure all of these have the version command?

Reply:
yep